src/main/java/org/veupathdb/service/eda/us/model/UserDataFactory.java

```diff
@@ -121,6 +122,234 @@ private String addSchema(String sqlConstant) {
+    DV_COL_DISPLAY_NAME  = "display_name",
+    DV_COL_DESCRIPTION   = "description",
+    DV_COL_PROVENANCE    = "provenance",
+    DV_COL_FUNCTION_NAME = "function_name",
+    DV_COL_CONFIG        = "config";
+
+  private static DerivedVariableRow resultSetToDVRow(ResultSet rs) {
+    return mapException(() -> new DerivedVariableRow(
+      rs.getString(DV_COL_VARIABLE_ID),
+      rs.getLong(DV_COL_USER_ID),
+      rs.getString(DV_COL_DATASET_ID),
+      rs.getString(DV_COL_ENTITY_ID),
+      rs.getString(DV_COL_DISPLAY_NAME),
+      rs.getString(DV_COL_DESCRIPTION),
+      mapIfPresent(rs.getString(DV_COL_PROVENANCE), raw -> new DerivedVariableProvenance(new JSONObject(raw))),
+      rs.getString(DV_COL_FUNCTION_NAME),
+      with(rs.getString(DV_COL_CONFIG), c -> mapException(() -> Utils.JSON.readTree(c), EXCEPTION_HANDLER))
+    ), EXCEPTION_HANDLER);
+  }
+
+  /**
+   * Prefix SQL for bulk selecting derived variables by ID.
+   *
+   * <p>
+   * Note the apostrophe at the end of the SQL prefix.
+   * </p>
+   */
+  // language=Oracle
+  private static final String BULK_SELECT_DERIVED_VARS_PREFIX =
+    "SELECT variable_id, user_id, dataset_id, entity_id, display_name, "
+      + "description, provenance, function_name, config FROM "
+      + TABLE_DERIVED_VARS + " WHERE variable_id IN ('";
+
+  /**
+   * Suffix SQL for bulk selecting derived variables by ID.
+   *
+   * <p>
+   * Note the apostrophe at the beginning of this SQL suffix.
+   * </p>
+   */
+  private static final String BULK_SELECT_DERIVED_VARS_SUFFIX = "')";
+
+  /**
+   * Returns a list of {@link DerivedVariableRow} instances matching the list of
+   * given input IDs.  If an input ID does not match a row in the database, no
+   * row for that ID will be returned.
+   *
+   * <p>
+   * <b>WARNING</b>: An SQL exception will be thrown if the given input list of
+   * IDs contains more than {@code 1000} elements.
+   * </p>
+   *
+   * @param ids List of derived variable IDs for the rows to look up.
+   * <br>
+   * THIS LIST MUST BE VALIDATED BEFOREHAND TO PREVENT SQL INJECTION!  Derived
+   * variable IDs must be UUID values.
+   *
+   * @return A list of {@link DerivedVariableRow} instances that were found to
+   * match one of the given input IDs.  This list will be at most the same size
+   * as the input list of IDs, but may be smaller.
+   */
+  public List<DerivedVariableRow> getDerivedVariables(Collection<String> ids) {
+    var sql = addSchema(ids.stream()
+      .collect(Collectors.joining("','", BULK_SELECT_DERIVED_VARS_PREFIX, BULK_SELECT_DERIVED_VARS_SUFFIX)));
+
+    LOG.debug("Bulk derived variable lookup for {} derived vars using the query {}", ids.size(), sql);
+
+    return mapException(() -> new SQLRunner(Resources.getUserDataSource(), sql)
+      .executeQuery(rs -> {
+        var out = new ArrayList<DerivedVariableRow>(ids.size());
+
+        while (rs.next())
+          out.add(resultSetToDVRow(rs));
+
+        return out;
```

Comment:
I think this should be checked by the calling method, if that calling method expects them to be the same length.  I can imagine situations where it would be expected that the sizes differ.  We use this practice in other services.  Though I will need to double check that the sizes are compared in the calling location in this case.